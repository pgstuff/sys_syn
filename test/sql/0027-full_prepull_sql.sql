BEGIN;

CREATE EXTENSION tinyint
    SCHEMA public;

CREATE EXTENSION sys_syn;

CREATE SCHEMA user_data
    AUTHORIZATION postgres;

CREATE TABLE user_data.test_table (
        test_table_id integer NOT NULL,
        test_table_text text,
        CONSTRAINT test_table_pid PRIMARY KEY (test_table_id));

INSERT INTO sys_syn.in_groups_def VALUES ('in');

SELECT sys_syn.pre_pull_add_sql('user_data.test_table'::regclass, 'in');

DO $$BEGIN
    EXECUTE sys_syn.pre_pull_add_sql('user_data.test_table'::regclass, 'in');
END$$;

ROLLBACK;
