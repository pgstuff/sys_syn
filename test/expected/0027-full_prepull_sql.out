BEGIN;
CREATE EXTENSION tinyint
    SCHEMA public;
CREATE EXTENSION sys_syn;
CREATE SCHEMA user_data
    AUTHORIZATION postgres;
CREATE TABLE user_data.test_table (
        test_table_id integer NOT NULL,
        test_table_text text,
        CONSTRAINT test_table_pid PRIMARY KEY (test_table_id));
INSERT INTO sys_syn.in_groups_def VALUES ('in');
SELECT sys_syn.pre_pull_add_sql('user_data.test_table'::regclass, 'in');
                                           pre_pull_add_sql                                            
-------------------------------------------------------------------------------------------------------
 CREATE UNLOGGED TABLE user_data.test_table_prepull_full (                                            +
         trans_id_in sys_syn.trans_id DEFAULT sys_syn.trans_id_get() NOT NULL,                        +
         test_table_id   integer,                                                                     +
         test_table_text text                                                                         +
 );                                                                                                   +
 INSERT INTO sys_syn.prepulls_def(                                                                    +
         prepull_id,     schema)                                                                      +
 VALUES ('test_table',   'user_data');                                                                +
 CREATE OR REPLACE FUNCTION user_data.test_table_prepull_full()                                       +
         RETURNS boolean AS                                                                           +
 $BODY$                                                                                               +
 DECLARE                                                                                              +
         _prepull_def            sys_syn.prepulls_def%ROWTYPE;                                        +
         _in_pull_state          sys_syn.in_pulls_state%ROWTYPE;                                      +
         _possible_changes       BOOLEAN := FALSE;                                                    +
 BEGIN                                                                                                +
         _prepull_def := (                                                                            +
                 SELECT  prepulls_def                                                                 +
                 FROM    sys_syn.prepulls_def                                                         +
                 WHERE   prepulls_def.prepull_id = 'test_table');                                     +
                                                                                                      +
         IF NOT pg_try_advisory_lock('sys_syn.prepulls_def'::regclass::int, _prepull_def.lock_id) THEN+
                 RAISE NOTICE 'Prepull % is running.', _prepull_def.prepull_id;                       +
                 RETURN FALSE;                                                                        +
         END IF;                                                                                      +
                                                                                                      +
         TRUNCATE user_data.test_table_prepull_full;                                                  +
                                                                                                      +
         PERFORM sys_syn.in_trans_pre_pull_start(FALSE);                                              +
                                                                                                      +
         INSERT INTO user_data.test_table_prepull_full (                                              +
                 trans_id_in,                                                                         +
                 test_table_id,                                                                       +
                 test_table_text)                                                                     +
         SELECT  sys_syn.trans_id_get(),                                                              +
                 test_table_id,                                                                       +
                 test_table_text                                                                      +
         FROM    user_data.test_table;                                                                +
         IF FOUND THEN _possible_changes := TRUE; END IF;                                             +
                                                                                                      +
         PERFORM sys_syn.in_trans_finish();                                                           +
                                                                                                      +
         -- Remove changes that are older than the full dataset.                                      +
         -- DELETE FROM test_table_prepull_changes WHERE trans_id_in < sys_syn.trans_id_get();        +
                                                                                                      +
         PERFORM pg_advisory_unlock('sys_syn.prepulls_def'::regclass::int, _prepull_def.lock_id);     +
                                                                                                      +
         RETURN _possible_changes;                                                                    +
 END;                                                                                                 +
 $BODY$                                                                                               +
   LANGUAGE plpgsql VOLATILE                                                                          +
   COST 2000;                                                                                         +
 SELECT sys_syn.in_table_add_sql(                                                                     +
         relation        => 'user_data.test_table_prepull_full'::regclass,                            +
         in_group_id     => 'in',                                                                     +
         schema          => 'user_data',                                                              +
         id_columns      => ARRAY['test_table_id'],                                                   +
         no_diff_columns => NULL,                                                                     +
         omit_columns    => ARRAY[]::TEXT[],                                                          +
         limit_to_columns=> NULL,                                                                     +
         full_prepull_id => 'test_table',                                                             +
         changes_prepull_id=> NULL,                                                                   +
         in_table_id     => 'test_table');                                                            +
 
(1 row)

DO $$BEGIN
    EXECUTE sys_syn.pre_pull_add_sql('user_data.test_table'::regclass, 'in');
END$$;
ROLLBACK;
